docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,166 @@
+  some external items are actually non-versioned, so are inherently floating.
+
+##### Expected behaviour regarding external references with floating tags
+
+Changes on external references is monitored and acted on (git branches,
+device images, secrets), the device controller should re-evaluate the device
+spec and apply the changes.
+
+Several mechanisms can be used to track external references:
+* Polling
+* Webhooks
+
+#### `/device/{name}/rendered` (device side)
+This endpoint is used by the device agent to retrieve the rendered device spec, this endpoint is
+read-only, each device can only access it's own device rendered spec based on the device
+certificate identification or an administrator (for debug purposes).
+
+The proposed changes include:
+* adding a version field, which identifies the version of the configuration.
+* rendering the config details in yaml format instead of a json blob (better for human readability while debugging)
+* the config field is a list of *inline* config items. Using a separate schema for this field which won't allow other config
+  types with external references.
+
+Behavior changes:
+* If a device configuration is not yet rendered for some reason, FC will queue the device for rendering and instruct
+  the device to try again later by returning 204 No Content.
+
+#### `/device/{name}/status` (device side)
+
+This endpoint is used by the device agent to report the status of the device, this endpoint is write-only,
+each device can only access it's own device status based on the device certificate identification.
+
+Reporting to this endpoint also serves as a heartbeat mechanism, the device agent should report the status,
+if a device stops reporting to this endpoint in the expected timelines will be considered offline.
+
+### Fleet API
+
+Fleet is an automation layer on top of Device API, the DeviceTemplate is a mechanism to control drift
+and enforce policy across devices in a fleet.
+
+A DeviceTemplate is a template for a device spec, it's a way to define group specific configuration.
+
+We are proposing extension mechanisms to this API to allow for more complex configurations where
+device labels or name would influence the configuration generated to the device.
+
+https://github.com/flightctl/flightctl/pull/248
+
+#### Expected behavior
+
+When fleet.speec.template.metadata or fleet.spec.template.spec is updated, the Flet deployment controller
+starts rolling out that update to devices. This update can take long (depending on the upgrade strategy)
+
+When a template is applied to a specific device via the Device API:
+* Floating tags for external references (os image, git refs, etc..) get converted to stable tags (hashes),
+  this gives flightctl the ability to check images being deployed on devices before the device could auto pick-up
+  a floating tag. (i.e. checks verifying the installation of the flightctl agent, and configuration)
```

Comment:
> This is too late IMO. We need to freeze at the fleet level, perform all of the checks there, and only then proceed to roll out to devices.

I guess you are referring to the case where:
1) We start rolling out devices, i.e. to image quay.io/myproject/image:latest pointing to aabbccddeeff
2) Some devices are frozen to aabbccddeeff
3) The :latest tag changes to 001122334455
4) Some devices will point to 001122334455

The idea here is that the devices should then slowly converge into the new version, when the controller detects that the hash has changed, the devices with the older tag will be updated to the newer one.


I believe that when customers start using floating tags, specially like this "latest" they must assume this type of issue (which would slowly converge in the desired state : all devices having the same version).

And if they want to avoid something like this, then they must use "tags" or less floating tags (per/version) ... or simply hashes. We should document and recommend here.

The same issue happens in kubernetes if you are deploying lots of pods.

Reply:
How I think it should work:
1. When quay.io/myproject/image:latest points to aabbccddeeff, FC calls it "v1", and starts rolling it out to all devices.
2. When it points to 001122334455, FC calls it "v2", and rolls it out to devices.

When someone sees a device running v1, they actually know what it's running.  It's also obvious how many devices are running aabbccddeeff, what the success rate is, and we can roll back to known-good hashes.

Otherwise, as currently described in the doc, FC is rolling out some spec.  The meaning behind it changes at some point.  The success rate suddenly drops.  There is no clear indication of what's going on because the fleet description didn't change.  So we roll back to some version that worked, but the fleet doesn't remember any hashes that worked - it only knows branches.

Reply:
What I am saying is that, that "uncertainty" must be admissible because the administrator is selecting a floating tag, which has the ability to float over time, and that we "do not recommend" (in docs we should do that).

We could go to the lengths of alleviating this problem, i.e. by pre-scanning all devices in a fleet and making a reference manifest (similar to the manifest openshift has for all the images)... that manifest would contain a list of all possible references on a fleet, and then when deploying we would use that instead of using dynamic resolution.

But, I think this is something we should not get into. we should just state: "This is a risk you are assuming if you use floating tags..." it could take more time to converge, you could find devices with different versions (as it happens in Kubernetes with floating tags and pods...)

Reply:
Let's keep variables as a separate topic.

Let's say a user supplies a tag or branch in the fleet specification, and the contents of that tag/branch change (either on purpose or accidentally), then the devices stop acting as a fleet.

The fleet controller simply copies the branch/tag to the specs of the devices.  There will be some observability and policy around how this is rolled out (batches, status, etc.).  Let's say the fleet was successfully updated.  Then suddenly, someone pushes to git, and the system breaks.  Devices start updating to the new hashes on their own, without obeying the fleet policies, and start failing.  None of this activity is visible on the fleet level.  There's no automatic rollback because this update was done outside of the fleet context.

I would say that this behavior is broken.

Reply:
> Let's keep variables as a separate topic.
> 
> Let's say a user supplies a tag or branch in the fleet specification, and the contents of that tag/branch change (either on purpose or accidentally), then the devices stop acting as a fleet.
> 
> The fleet controller simply copies the branch/tag to the specs of the devices. There will be some observability and policy around how this is rolled out (batches, status, etc.). Let's say the fleet was successfully updated. Then suddenly, someone pushes to git, and the system breaks. Devices start updating to the new hashes on their own, without obeying the fleet policies, and start failing. None of this activity is visible on the fleet level. There's no automatic rollback because this update was done outside of the fleet context.

Ok, I suspect I didn't do my best to explain this on the RFE:

This RFE says (or I tried to say that) that the fleet controller freezes the tag in the device.

However, we allow devices to be populated with floating tags when a fleet does not manage them (i.e. a external system).


> I would say that this behavior is broken.

Reply:
Resolving based on our discussion here, but feel free to re-open https://redhat-internal.slack.com/archives/C01KER8QV33/p1715593490026039